state/clickhouse/clickhouse.go

```diff
@@ -0,0 +1,435 @@
+	//nolint:gosec
+	insertQuery := `
+		INSERT INTO ` + c.config.DatabaseName + `.` + c.config.TableName + ` (key, value, etag, expire)
+		VALUES (?, ?, ?, ?)
+	`
+
+	_, err = c.db.ExecContext(ctx, insertQuery, req.Key, value, etag, expireTime)
+	if err != nil {
+		// If the key exists, update it
+		// Note: Database and table names are validated during metadata parsing
+		// and come from trusted configuration, so direct string concatenation is acceptable here
+		//nolint:gosec
+		updateQuery := `
+			ALTER TABLE ` + c.config.DatabaseName + `.` + c.config.TableName + ` 
+			UPDATE value = ?, etag = ?, expire = ?
+			WHERE key = ?
+		`
+
+		_, updateErr := c.db.ExecContext(ctx, updateQuery, value, etag, expireTime, req.Key)
+		if updateErr != nil {
+			return fmt.Errorf("error updating value: %v", updateErr)
+		}
+	}
+
+	return nil
+}
+
+func (c *StateStore) Delete(ctx context.Context, req *state.DeleteRequest) error {
+	if req.Key == "" {
+		return errors.New("key is empty")
+	}
+
+	// Handle ETag for optimistic concurrency
+	if req.ETag != nil && *req.ETag != "" {
+		// First, get the current etag
+		currentETag, err := c.getETag(ctx, req.Key)
+		if err != nil {
+			return err
+		}
+
+		// If an etag exists and it doesn't match the provided etag, return error
+		if currentETag != "" && currentETag != *req.ETag {
+			return state.NewETagError(state.ETagMismatch, nil)
+		}
+	}
+
+	// Note: Database and table names are validated during metadata parsing
+	// and come from trusted configuration, so direct string concatenation is acceptable here
+	//nolint:gosec
+	query := "DELETE FROM " + c.config.DatabaseName + "." + c.config.TableName + " WHERE key = ?"
+	_, err := c.db.ExecContext(ctx, query, req.Key)
+	return err
+}
+
+func (c *StateStore) marshal(v any) (string, error) {
+	var value string
+	switch v := v.(type) {
+	case []byte:
+		value = string(v)
+	case string:
+		value = v
+	default:
+		bt, err := utils.Marshal(v, json.Marshal)
+		if err != nil {
+			return "", err
+		}
+		value = string(bt)
+	}
+	return value, nil
+}
+
+func parseTTL(metadata map[string]string) (int, error) {
+	if metadata == nil {
+		return 0, nil
+	}
+	ttl, ok := metadata["ttlInSeconds"]
+	if !ok || ttl == "" {
+		return 0, nil
+	}
+
+	ttlMetadata := map[string]string{
+		"ttlInSeconds": ttl,
+	}
+
+	ttlPtr, err := utils.ParseTTL(ttlMetadata)
+	if err != nil {
+		return 0, fmt.Errorf("error parsing TTL: %v", err)
+	}
+
+	if ttlPtr == nil {
+		return 0, nil
+	}
+
+	return *ttlPtr, nil
+}
+
+func parseAndValidateMetadata(metadata state.Metadata) (clickhouseMetadata, error) {
+	config := clickhouseMetadata{}
+
+	err := kitmd.DecodeMetadata(metadata.Properties, &config)
+	if err != nil {
+		return config, err
+	}
+
+	if config.ClickhouseURL == "" {
+		return config, errors.New("ClickHouse URL is missing")
+	}
+
+	if config.DatabaseName == "" {
+		return config, errors.New("ClickHouse database name is missing")
+	}
+
+	if config.TableName == "" {
```

Comment:
are there fixed char limits on these we should check for by chance?

Reply:
@sicoyle [looks like it's here](https://clickhouse.com/docs/sql-reference/functions/other-functions#getMaxTableNameLengthForDatabase).
@middt can you add a length check as well?